### ISSUE TITLE
Final_Model not work

### ISSUE THREAD

**Issue by yzl96** [camigord/ML_CapstoneProject#1]
Hi,camigord. I'm recurrenting your model, there are something I don't understand about you Regression Model output , you say Output[0:1] means confidence score ,why there are two numbers , I really don't understand .
In 7_Evaluate_Models.ipynb:
if prediction[0][1]>=0.7:
    # Crop the predicted bounding box
    im2 = clone.crop((prediction[0][3], prediction[0][2], prediction[0][5], prediction[0][4])).resize([54,54], Image.ANTIALIAS)

How did you get the 0.7.

Thank you very much

**Comment by camigord** [comment]
Hi @yzl96,

It is a long time ago since I wrote that code and I need to apologize for how disorganized everything looks like...

About your question: the "regression model" actually operates both as a classification model (to classify whether there is a number in the image or not) and as a regression model (to identify the bounding box). For this reason, the regression model output actually looks like this:

```
[no_number, is_number, top, left, bottom, right]
```

So, the first two values are used to classify the image as containing a number or not. This can be seen in the declaration of the weights _presence-w_ and _presence-b_, where you can see that I am defining an output of two values:

```
    # Classifier output: whether there is a number or not
    presence_w = tf.get_variable("WO1", shape=[num_hidden1, 2],\
                           initializer=tf.contrib.layers.xavier_initializer())
    presence_b = tf.Variable(tf.constant(1.0, shape=[2]), name='BO1')
```

So, there is no "confidence score" as such, but I use the likelihood of the class "is_number" as a confidence value to determine the presence of a number (prediction[0][1]). The value 0.7 is just a threshold which can "more or less" be understood as having 70% confidence on the classification.

I hope that solves your question and do not hesitate to post any other comment you may have.

Regards,

**Comment by yzl96** [comment]
Hi @camigord：
    Thank you very much for responsing so quickly.
    Now I know what you mean,but there are still something I don't know.
    I see the architecture of your Regression Model , I don't know what's the difference of no_number and is_number in the architecture of you Model, or it depends on you train_label, maybe the label looks like that: [ *,is_number,*,* ], is_number located in label[1] . so you think prediction[1] means is_number.

**Comment by camigord** [comment]
Hi @yzl96 ,

I am not sure I fully understood your question. So I will try to summarize some things:

The output of the regression model looks like I said before:
```
[no_number, is_number, top, left, bottom, right]
```

The labels used for training, as you can see in the file "6_Train_Regression_Model.ipynb", contain 5 numbers. The last four correspond to the bounding box while the first one (label[0]) represents whether the image contains a number (1) or not (0). 

When I trained the network (also in file "6_Train_Regression_Model.ipynb") I defined the loss as:
```
    # Loss
    loss_presence = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(Outputs[0], tf_train_labels[:,0]))
    l2diff = tf.sqrt(tf.reduce_sum(tf.square(tf.sub(tf.to_float(tf_train_labels[:,1:]), Outputs[1])),reduction_indices=1))
    loss_bbox = tf.reduce_mean(l2diff)
    total_loss = loss_presence + loss_bbox
```

where _Outputs[0]_ is our _classifier_ output [no_number, is_number] and _Outputs[1]_ is the bounding box. 

I think you may be confused about the fact that one single label in our training data is used to train the two values output of the classifier part, and if that is the case, you should take a look at the function __tf.nn.sparse_softmax_cross_entropy_with_logits()__ which takes care of training the classifier output.

**Comment by yzl96** [comment]
Hi, @camigord,
   Now I understood. I confused the function:
   tf.nn.softmax_cross_entropy_with_logits(labels=None, logits=None)    and  tf.nn.sparse_softmax_cross_entropy_with_logits(labels=None, logits=None)
   In tf.nn.sparse_softmax_cross_entropy_with_logits(labels=None, logits=None) , logits looks like that:
[batch_size, num_classes] and labels looks like that [batch_size]  
for example , if detect number, label[0] = 1 , on the contrary, label[0] = 0 , so one single label in our training data is used to train the two values output of the classifier part.
Thank you very much ,your code is very cool.
